File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;

class Solution {
    Set<Integer> primeSet = new HashSet<>();

    public int solution(String numbers) {
        this.makeWord("", numbers);
        return primeSet.size();
    }

    void makeWord(String word, String others){
        if(!word.equals("")){
            int n = Integer.parseInt(word);
            if(isPrime(n)) primeSet.add(n);
        }

        for(int i = 0; i < others.length(); i++){
            makeWord(word + others.charAt(i), others.substring(0, i) + others.substring(i+1));
        }
    }

    Boolean isPrime(int n){
        if(n < 2) return false;
        int limit = (int) Math.sqrt(n);
        for(int i=2; i <= limit; i++)
            if(n % i == 0) return false;
        return true;
    }
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.kt
import kotlin.math.sqrt
    
class Solution {
    private lateinit var visited: BooleanArray
    private lateinit var numbers: String 
    private val answer = HashSet<Int>()
    
    fun DFS(word: String){
        if(word.isNotBlank()){
            val wordToInt = word.toInt()
            if(isPrime(wordToInt))
                answer.add(wordToInt)
        }
        
        for(i in numbers.indices){
            if(!visited[i]){
                visited[i] = true
                DFS(word + numbers[i])
                visited[i] = false
            }
        }
    }

    fun solution(numbers: String): Int {
        visited = BooleanArray(numbers.length + 1)
        this.numbers = numbers
        DFS("")
        return answer.size;
    }

    fun isPrime(n: Int): Boolean{
        if (n < 2) return false
        for (i in 2.. sqrt(n.toDouble()).toInt()){
            if (n % i == 0) return false
        }
        return true
    }
}